packages/vm/lib/buildBlock.ts

```diff
@@ -0,0 +1,241 @@
+      gasLimit: opts.headerData?.gasLimit ?? opts.parentBlock.header.gasLimit,
+    }
+  }
+
+  /**
+   * Throws if the block has already been built or reverted.
+   */
+  checkStatus() {
+    if (this.built) {
+      throw new Error('Block has already been built')
+    }
+    if (this.reverted) {
+      throw new Error('State has already been reverted')
+    }
+  }
+
+  /**
+   * Returns the cumulative gas used by the transactions added to the block.
+   */
+  gasUsed(): BN {
+    return this.transactionResults.reduce(
+      (accumulator: BN, result: RunTxResult) => accumulator.iadd(result.gasUsed),
+      new BN(0)
+    )
+  }
+
+  /**
+   * Calculates and returns the transactionsTrie for the block.
+   */
+  private async transactionsTrie() {
+    const trie = new Trie()
+    for (const [i, tx] of this.transactions.entries()) {
+      await trie.put(encode(i), tx.serialize())
+    }
+    return trie.root
+  }
+
+  /**
+   * Calculates and returns the logs bloom for the block.
+   */
+  private bloom() {
+    const bloom = new Bloom()
+    for (const txResult of this.transactionResults) {
+      // Combine blooms via bitwise OR
+      bloom.or(txResult.bloom)
+    }
+    return bloom.bitvector
+  }
+
+  /**
+   * Calculates and returns the receiptTrie for the block.
+   */
+  private async receiptTrie() {
+    const gasUsed = new BN(0)
+    const receiptTrie = new Trie()
+    for (const [i, txResult] of this.transactionResults.entries()) {
+      const tx = this.transactions[i]
+      gasUsed.iadd(txResult.gasUsed)
+      const { encodedReceipt } = await generateTxReceipt.bind(this.vm)(tx, txResult, gasUsed)
+      await receiptTrie.put(encode(i), encodedReceipt)
+    }
+    return receiptTrie.root
+  }
+
+  /**
+   * Adds the block miner reward to the coinbase account.
+   */
+  private async rewardMiner() {
+    const minerReward = new BN(this.vm._common.param('pow', 'minerReward'))
+    const reward = calculateMinerReward(minerReward, 0)
+    const coinbase = this.headerData.coinbase
+      ? new Address(toBuffer(this.headerData.coinbase))
+      : Address.zero()
+    await rewardAccount(this.vm.stateManager, coinbase, reward)
+  }
+
+  /**
+   * Run and add a transaction to the block being built.
+   * Please note that this modifies the state of the VM.
+   * Throws if the transaction's gasLimit is greater than
+   * the remaining gas in the block.
+   */
+  async addTransaction(tx: TypedTransaction) {
+    this.checkStatus()
+
+    if (!this.checkpointed) {
+      await this.vm.stateManager.checkpoint()
+      this.checkpointed = true
+    }
+
+    // According to the Yellow Paper, a transaction's gas limit
+    // cannot be greater than the remaining gas in the block
+    const blockGasLimit = new BN(toBuffer(this.headerData.gasLimit))
+    const blockGasRemaining = blockGasLimit.sub(this.gasUsed())
+    if (tx.gasLimit.gt(blockGasRemaining)) {
+      throw new Error('tx has a higher gas limit than the remaining gas in the block')
```

Comment:
If this throws, then we should revert the state manager checkpoint right?

Reply:
as discussed in the call, I think this is okay since the consumer of the API can decide if the tx doesn't fit if they would rather add another one, or build to finish the block. if they want to revert the whole block they can use `revert()`.